### ISSUE TITLE
JSON fails to parse when there is only a nested map with no non-nested fields

### ISSUE THREAD

**Issue by anuraaga** [hashicorp/terraform#18330]
I am trying to use a JSON config file but when a resource only has a nested map or two (`kubernetes_namespace` only has `metadata`, `kubernetes_secret` only has `metadata` and `data`), it fails to parse with an error `resource repeated multiple times`. If I add a no-op singular field as well, like `"count": 1`, it works.

### Terraform Version

```
Terraform v0.11.7
```

#### Not working

provider.tf (doesn't get past parse so these values don't matter)
```hcl
provider "google" {
  project = "project1"
  region = "asia-northeast1"
}

provider "kubernetes" {}
```

test.tf.json
```json
{
  "resource" : {
    "kubernetes_namespace" : {
      "namespace1" : {
        "metadata" : {
          "name" : "namespace1"
        }
      }
    },
    "kubernetes_secret" : {
      "secret1" : {
        "metadata" : {
          "name" : "secret1",
          "namespace" : "namespace1"
        },
        "data" : {
          "service-account.json" : "contents"
        }
      }
    }
  }
}
```

#### Working
```hcl
provider "google" {
  project = "project1"
  region = "asia-northeast1"
}

provider "kubernetes" {}
```

test.tf.json
```json
{
  "resource" : {
    "kubernetes_namespace" : {
      "namespace1" : {
        "count": 1,
        "metadata" : {
          "name" : "namespace1"
        }
      }
    },
    "kubernetes_secret" : {
      "secret1" : {
        "count": 1,
        "metadata" : {
          "name" : "secret1",
          "namespace" : "namespace1"
        },
        "data" : {
          "service-account.json" : "contents"
        }
      }
    }
  }
}
```

### Debug Output

```
2018/06/26 21:23:30 [INFO] Terraform version: 0.11.7  41e50bd32a8825a84535e353c3674af8ce799161
2018/06/26 21:23:30 [INFO] Go runtime version: go1.10.1
2018/06/26 21:23:30 [INFO] CLI args: []string{"/home-unencrypted/anuraag/.gradle/curiostack/terraform/0.11.7/terraform", "plan"}
2018/06/26 21:23:30 [DEBUG] Attempting to open CLI config file: /home/anuraag/.terraformrc
2018/06/26 21:23:30 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2018/06/26 21:23:30 [INFO] CLI command args: []string{"plan"}

2018/06/26 21:23:30 [DEBUG] plugin: waiting for all plugin processes to complete...
Error: kubernetes_secret.secret1: resource repeated multiple times
```

### Expected Behavior

Initializes properly

### Actual Behavior

There is an error

### Steps to Reproduce

1. `terraform init`

**Comment by apparentlymart** [comment]
Hi @anuraaga! Sorry for this very strange behavior and thanks for reporting it.

The next major release of Terraform will include a new approach to JSON parsing that should avoid strange misinterpretations like this: the new JSON configuration parser has access to the expected configuration schema which it then uses to resolve the ambiguities that naturally arise when attempting to extract the higher-level block structure from JSON.

I believe this particular issue should already be fixed in our development branch, but there is other work in progress that we'll need to complete before we can test this fully and so we'll use this issue as a reminder to test that this is indeed fixed during our testing phase prior to the release.

For now, your workaround of using `count` here to force it is a clever way to trick the current parser, although it will unfortunately have an unintended side-effect in the next major release where the _presence_ of `count` (even if set to 1) will cause the expression `kubernetes_secret.secret1` to return a list of objects rather than the single object it'd return when `count` isn't set at all. Expressions using attributes of this would need to be written as e.g. `kubernetes_secret.secret1[0].id`. This is a consequence of the new capability to use an entire resource as a value within expressions, rather than just attributes of resources.

If you'd like to make your configuration ready for the new release now, I think you could write this as `kubernetes_secret.secret1.0.id` to get something that would work for both the current release and the next release. (The next release will treat `kubernetes_secret.secret1.0` as a compatibility alias for `kubernetes_secret.secret1[0]`).

Thanks again for reporting this!

**Comment by anuraaga** [comment]
Thanks for the note! I changed `count: 1` to `depends_on: []` since it sounds like the latter would be forward-compatible without the weirdness. Glad to hear about it in advance :)

**Comment by apparentlymart** [comment]
Hi again, @anuraaga! Sorry for the long silence here.

The changes I mentioned before have now been merged into master in preparation for the forthcoming v0.12.0 release.

Unfortunately we didn't include the `kubernetes` provider in our alpha1 or alpha2 prerelease builds (just to reduce the overhead of preparing these non-upstream provider builds for the alpha) so I can't verify your configuration exactly as given, but I did find a resource in the AWS provider that also has two map arguments that I could use as a proxy:

```json
{
    "resource": {
        "aws_cognito_identity_provider": {
            "example": {
                "attribute_mapping": {
                    "a": "b"
                },
                "provider_details": {
                    "a": "b"
                }
            }
        }
    }
}
```

Planning this _does_ given an error, but since it's a validation error within the resource itself I can tell that the top-level configuration block decoding succeeded, where in v0.11 and prior it was failing:

```
$ terraform plan

Error: Missing required argument

  on json-map-attr.tf.json line 11, in resource.aws_cognito_identity_provider.example:
  11:             }

The argument "provider_name" is required, but no definition was found.


Error: Missing required argument

  on json-map-attr.tf.json line 11, in resource.aws_cognito_identity_provider.example:
  11:             }

The argument "provider_type" is required, but no definition was found.


Error: Missing required argument

  on json-map-attr.tf.json line 11, in resource.aws_cognito_identity_provider.example:
  11:             }

The argument "user_pool_id" is required, but no definition was found.

```

Since this is fixed in master, I'm going to close this out now. The fix will be included in the v0.12.0 final release. Thanks for reporting this, and thanks for your patience while we got it fixed.
